### PR TITLE
Feature/dynamic plot zoom

### PR DESCRIPTION
**Pull request by timtmok**
Address #4210 

Reuses the component created for static plots to enable pan and zoom. Modified the component to recalculate the image scaling so that the image is properly displayed when switching between plots.

Fixed a warning about multi-window in the dynamic plot component. It might have been a problem in the future with multi-window instances. Also, fixed a refresh problem with the component re-rendering unnecessarily due to missing hook dependencies.

image

```R
library(datasets)
hist(iris$Sepal.Width)
```
### QA Notes
An area to focus on is changing the plot. The zoom and sizing policy should update after changing plots if they were changed while viewing the previous plot.

Smoke tests were updated to find the plots since the structure changed to encapsulate the image in a scrollable element. Also changed the check on the zoom option to ensure it is visible with a dynamic plot.

### CODE REVIEW

**Comment by timtmok** [comment]
The smoke tests will need some updates. The image diffs show that it's scaled to the wrong aspect ratio. I think the image is taken too soon before the image is scaled properly. Trying the change out locally shows that the image looks good but there might be an initial render at the wrong scale.

**Comment by nstrayer** [comment]
Do you have any code for a quick repro of what's been done here?

**Comment by timtmok** [comment]
> Do you have any code for a quick repro of what's been done here?

@nstrayer I've updated the description with the code that generated the plot in the screenshot.

I'm working on updating the smoke tests to handle the image comparison failure. It looks like the captured image is taken before the plot is adjusted for proper aspect ratio.

**Review by nstrayer — COMMENTED** [review]
Testing with 
```python
import matplotlib.pyplot as plt

# x axis values
x = [1,2,3]
# corresponding y axis values
y = [2,4,1]

# plotting the points
plt.plot(x, y)

# naming the x axis
plt.xlabel('x - axis')
# naming the y axis
plt.ylabel('y - axis')

# giving a title to my graph
plt.title('My first graph!')

# function to show the plot
plt.show()
```
As pulled from the smoke test. 

Not sure if this is the desired behavior but I think it's confusing that "fill" and "50%" are the same and then 75% is a zoom in. Is this the desired behavior or due to quirks in retina displays? 
image

File: src/vs/workbench/contrib/positronPlots/browser/components/panZoomImage.tsx [hunk 1]
```diff
@@ -23,17 +23,17 @@ interface PanZoomImageProps {
  * @returns The rendered component.
  */
 export const PanZoomImage = (props: PanZoomImageProps) => {
-	const [width, setWidth] = React.useState<number>(1);
-	const [height, setHeight] = React.useState<number>(1);
+	const [naturalWidth, setNaturalWidth] = React.useState<number>(1);
+	const [naturalHeight, setNaturalHeight] = React.useState<number>(1);
+	const [scrollableWidth, setScrollableWidth] = React.useState<number>(1);
+	const [scrollableHeight, setScrollableHeight] = React.useState<number>(1);
```
[the remark this answers is not in the document]
Comment (timtmok) [reply src/vs/workbench/contrib/positronPlots/browser/components/panZoomImage.tsx:29]: I haven't noticed a flash. Maybe a better initial value here would prevent React from an unnecessary render?

**Comment by timtmok** [comment]
> Not sure if this is the desired behavior but I think it's confusing that "fill" and "50%" are the same and then 75% is a zoom in. Is this the desired behavior or due to quirks in retina displays?

It's a scaling percentage on the image size. "Fill" is a content fit and may match one of the percentage levels depending on the image size.

Maybe it makes more sense to use preset names like "0.5x" instead of "50%"? And "Fill" should be "Fit" or "Content Fit"?

**Comment by timtmok** [comment]
I'm noticing that the image comparison diffs from test run artifacts may look like the wrong aspect ratio. The diff may show a skewed image if the compared images are different resolutions.

The middle image is the captured image from the test from a local run. It looks fine during the test run but is a different aspect ratio than the reference image. `resemblejs` seems to show a diff where the resolutions are the same (probably to make it easier to see the differences).
graphvizDiff

So, I'm seeing something similar in the diff from the CI run but this likely means generating a new reference image is all that's needed.

File: src/vs/workbench/contrib/positronPlots/browser/components/panZoomImage.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (nstrayer) [reply src/vs/workbench/contrib/positronPlots/browser/components/panZoomImage.tsx:29]: I'd lean towards something that isn't a valid value, just for clarity sake. Like a negative 1 or `null` or `undefined`.

**Review by nstrayer — DISMISSED** [review]
Code is looking good! 

I did notice that there's some not-perfect lining up of scrollableness in the fill situation. Aka when you're at the `fill` setting the size of the image is a couple pixels larger than the view area and thus you get scroll behavior. I'm not sure if this is intrinsic to this PR or the `DomScrollableElement` item itself. 

If you think it's not applicable to this PR feel free to merge and I can create another issue about it. 

I also have some reservations about the tests needing as large of a timeout as used, but I'll defer judgment on that to QA. 

https://github.com/user-attachments/assets/8bcebd80-4ea0-4e66-9e5d-a9e43a3ec8f7

File: test/smoke/src/areas/positron/plots/plots.test.ts [hunk 2]
```diff
@@ -80,12 +80,15 @@ plt.show()`;
 
 				await app.workbench.positronPlots.waitForCurrentPlot();
 
+				// give time for the React component to apply the zoom
+				await app.code.wait(3000);
```
[the remark this answers is not in the document]
Comment (timtmok) [reply test/smoke/src/areas/positron/plots/plots.test.ts:84]: I think I can remove this as I initially thought that CI was capturing the plot before React rendering was complete. I'm pretty sure it's the reference image that needs updating and that the diff image in the test artifacts was confusing to interpret.

**Comment by timtmok** [comment]
The smoke test was failing because the plots view actions weren't all visible. The extra button pushed the clear plots button just out of view. Resizing the viewport should fix it after generating some new reference images.

**Comment by timtmok** [comment]
Finally got the right reference images for the smoke tests. They'll generate new images as well if the image comparison fails.

The scrolling bug has been fixed as well. The image and container heights seemed to be correct but removing the `overflow: hidden` property fixed the issue.

**Review by nstrayer — DISMISSED** [review]
Looking good! 
I would pay attention to #4323 because there's some overlap in what is touched so depending on which PR is merged first there may be some pain dealing with conflicts.

**Pull request merged**